network: include ssid in ipv6 stable prefix address generation

rpigott opened the pull request:
networkd currently calculates stable prefix addresses exactly as described in RFC7217 and as recorded in a comment:

https://github.com/systemd/systemd/blob/0e3986bc1c3447aec1e9fe774a181d60677e686b/src/network/networkd-address-generation.c#L108-L109

The rfc has the following to say about the Network_ID parameter:

> Network_ID:
> Some network-specific data that identifies the subnet to which this interface is attached -- for example, the IEEE 802.11 Service Set Identifier (SSID) corresponding to the network to which this interface is associated.  Additionally, Simple DNA [RFC6059] describes ideas that could be leveraged to generate a Network_ID parameter.  This parameter is OPTIONAL.

Network_ID is an optional parameter which is currently excluded. Well I think it sounds like a good idea, particularly for achieving the stated privacy goals of stable prefix addresses on mobile wireless devices. The idea is to avoid using an identifying address on multiple networks.

Anyway hopefully this is all that is needed. I'm not sure if there is any expectation of address stability between systemd versions, nor any particular reasons we should wish to exclude this datum from the generation.

DaanDeMeyer commented on src/network/networkd-address-generation.c:
```diff
@@ -117,6 +117,8 @@ static void generate_stable_private_address_one(
                 siphash24_compress(&link->hw_addr.infiniband[INFINIBAND_ALEN - 8], 8, &state);
         else
                 siphash24_compress(link->hw_addr.bytes, link->hw_addr.length, &state);
+        if (link->ssid)
+                siphash24_compress_string(link->ssid, &state);
```
Please add a newline before and after the added logic

rpigott replied:
Ok, added them.

keszybz commented:
Please move the patch description to the patch. Right now it's just a text on the pull request which will not be attached to the commit.

keszybz commented:
> Network_ID is an optional parameter which is currently excluded. Well I think it sounds like a good idea,

Please insert something like "including it" in the second sentence… Right now it has the opposite meaning from the intended, IIUC.

---

I think this is reasonable. We like to keep stability, so we wouldn't introduce a change without a good reason, but here we have a good reason. This will need to be mentioned in NEWS to give people a heads-up.

yuwata reviewed (approved):
The code change itself is LGTM.
Please update the commit message and NEWS file.

rpigott commented:
Updated the commit message with the justification and updated NEWS to mention the change. For clarity it calls out that current users of prefixstable with wireless networks will observe a change.
